crates/services/gas_price_service/src/lib.rs

```diff
@@ -0,0 +1,206 @@
+use async_trait::async_trait;
+use fuel_core_services::{
+    RunnableService,
+    RunnableTask,
+    StateWatcher,
+};
+use fuel_core_types::fuel_types::BlockHeight;
+use std::sync::{
+    Arc,
+    RwLock,
+};
+
+pub mod static_updater;
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("Failed to update gas price algorithm")]
+    UpdateFailed(String),
+}
+
+pub struct GasPriceService<A, U> {
+    next_block_algorithm: Arc<RwLock<(BlockHeight, A)>>,
+    update_algorithm: U,
+}
+
+impl<A, U> GasPriceService<A, U>
+where
+    U: UpdateAlgorithm<Algorithm = A>,
+{
+    pub fn new(starting_block_height: BlockHeight, update_algorithm: U) -> Self {
+        let algorithm = update_algorithm.start(starting_block_height);
+        let next_block_algorithm =
+            Arc::new(RwLock::new((starting_block_height, algorithm)));
+        Self {
+            next_block_algorithm,
+            update_algorithm,
+        }
+    }
+
+    pub fn next_block_algorithm(&self) -> Arc<RwLock<(BlockHeight, A)>> {
+        self.next_block_algorithm.clone()
+    }
+}
+
+#[async_trait]
+pub trait UpdateAlgorithm {
+    type Algorithm;
+
+    fn start(&self, for_block: BlockHeight) -> Self::Algorithm;
+    async fn next(&mut self, for_block: BlockHeight) -> Self::Algorithm;
+}
```

Comment:
I think the algorithm itself should not be dependent on the block height. It has an internal state and `next` and `update` functions to iterate over this state.

The handling of the `for_block` should be done on the service level, since `service` knows what is the current block height and when the next block is imported.

Reply:
How does the producer know that the gas price service is on the same block as itself though? I feel like there needs to be a check somewhere that the producer is getting the algorithm for the block it's building.

Reply:
I've removed it for a separate reason. It works right now, but we might need to reconsider later.